packages/apollo-gateway/src/index.ts

```diff
@@ -343,40 +287,87 @@ export class ApolloGateway implements GraphQLService {
     };
   }
 
-  private startPollingServices() {
-    if (this.pollingTimer) clearInterval(this.pollingTimer);
+  private async updateComposition(options?: {
+    engine?: GraphQLServiceEngineConfig;
+  }) {
+    const previousSchema = this.schema;
+    const previousServiceDefinitions = this.serviceDefinitions;
+    const previousCompositionMetadata = this.compositionMetadata;
 
-    this.pollingTimer = setInterval(async () => {
-      let result: Await<ReturnType<Experimental_UpdateServiceDefinitions>>;
-      try {
-        result = await this.updateServiceDefinitions(this.config);
-      } catch (e) {
-        this.logger.debug(
-          'Error checking for schema updates. Falling back to existing schema.',
-          e,
-        );
-        return;
-      }
+    if (options && options.engine) {
+      if (!options.engine.graphVariant)
+        console.warn('No graph variant provided. Defaulting to `current`.');
+      this.engineConfig = options.engine;
+    }
 
-      if (!result.isNewSchema) {
-        this.logger.debug('No changes to gateway config');
-        return;
-      }
+    let result: Await<ReturnType<Experimental_UpdateServiceDefinitions>>;
+    try {
+      this.logger.debug('Loading configuration for gateway');
+      result = await this.updateServiceDefinitions(this.config);
+    } catch (e) {
+      this.logger.warn(
+        'Error checking for schema updates. Falling back to existing schema.',
+        e,
+      );
+      return;
+    }
 
-      if (this.queryPlanStore) this.queryPlanStore.flush();
-      this.logger.debug('Gateway config has changed, updating schema');
+    if (!result.isNewSchema) {
+      this.logger.debug('No changes to gateway config');
+    }
 
-      this.schema = this.createSchema(result.serviceDefinitions);
-      try {
-        this.onSchemaChangeListeners.forEach(listener =>
-          listener(this.schema!),
-        );
-      } catch (e) {
-        this.logger.debug(
-          'Error notifying schema change listener of update to schema.',
-          e,
-        );
-      }
+    if (
+      JSON.stringify(this.serviceDefinitions) ===
```

Comment:
Is this stringify stable? Should we `JSON.parse(JSON.stringify(...))`

Reply:
I think this is out of scope but worth demonstrating if this might be a concern in a test. The alternative is a deepEqual over objects containing `DocumentNode`s which would be preferable to avoid if we can.